File: Hearthstone-Arena-Card-Selector/src/Library.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;

public class Library {
	
	public Library() {
		
	}

	/**
	 * Parses CSV file. Returns a database with key:Pair, value:ListOfWins.
	 * @param file: formatted CSV file containing arena data.
	 */
	public HashMap<Pair, ArrayList<Integer>> parseFile(File file) 
	{
		HashMap<Pair, ArrayList<Integer>> ret = new HashMap<>();
		
		FileReader fr = null;
		try {
			fr = new FileReader(file);
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		}
		
		BufferedReader br = new BufferedReader(fr);
		
		try {
			int read_state = 0; //0: parsing header, 1: parsing content
			String arena_being_read_in = "";
			int arena_being_read_in_wins = 0;
			ArrayList<String> list_of_cards_in_this_arena = new ArrayList<>();
			while(br.ready()) 
			{
				String line = br.readLine();
				
				if(read_state == 0 && line.startsWith("game_version")) 
					read_state = 1;
				else if(read_state == 1)
				{
					//--if end of file
					if(line.startsWith(",") || line.startsWith(" "))
						break;
					
					String[] cols = line.split(",");	//game_version,end_time,date,character,card_title,wins,player_id
					
					//--if start of arena i
					if(!arena_being_read_in.equals(cols[1]+cols[2]+cols[6])) 
					{
						//--do stuff with info from arena i-1
						ArrayList<Pair> arena_being_read_in_pairs = makeUniquePairs(list_of_cards_in_this_arena, arena_being_read_in_wins);
						for(int i = 0; i < arena_being_read_in_pairs.size(); i++)
						{
							Pair currPair = arena_being_read_in_pairs.get(i);
							if(ret.containsKey(currPair))
							{
								ArrayList<Integer> win_history = ret.get(currPair);
								win_history.add(currPair.value);
								ret.put(currPair, win_history);
							}
							else
							{
								ArrayList<Integer> win_history = new ArrayList<>();
								win_history.add(currPair.value);
								ret.put(currPair, win_history);
							}
						}
						System.out.println("[Library] Arena '" + arena_being_read_in + "' has been added. (" + ret.size() + " in db)");
		
						//--prepare for arena i, like reset id and add character
						arena_being_read_in = cols[1]+cols[2]+cols[6];
						arena_being_read_in_wins = Integer.parseInt(cols[5]);
						list_of_cards_in_this_arena = new ArrayList<>();
						list_of_cards_in_this_arena.add(cols[3]);
					}
					list_of_cards_in_this_arena.add(cols[4]);
					
				}
			}
			
		} catch (IOException e) {
			e.printStackTrace();
		}
		
		return ret;
	}

	/**
	 * Gets metric of how well the pairing performs.
	 * @param map: the database used.
	 * @param pair: the pairing.
	 */
	public double getScore(HashMap<Pair, ArrayList<Integer>> map, Pair pair) {
		
		ArrayList<Integer> record = map.get(pair);
		Iterator<Integer> it = record.iterator();
		int sum = 0;
		while(it.hasNext())
		{
			sum += it.next();
		}
		
		return (double)sum/(double)record.size();
	}
	
	/**
	 * Gets a list
	 * @param db
	 * @param history
	 * @return
	 */
	public HashMap<String, Double> getAllNextScores(HashMap<Pair, ArrayList<Integer>> db, ArrayList<String> history)
	{
		
		return null;
	}
	
	
	private ArrayList<Pair> makeUniquePairs(ArrayList<String> list, int wins) 
	{
		ArrayList<Pair> ret = new ArrayList<>();
		
		for(int i = 0; i < list.size(); i++)
		{
			for(int j = 0; j < list.size(); j++)
			{
				if(i != j)
				{
					Pair currPair = new Pair(list.get(i), list.get(j), wins);
					if(!ret.contains(currPair)) ret.add(currPair);
				}
			}
		}
		
		return ret;
	}
	
}

File: Library.php
<?php
	require_once "Pair.php";

	class Library
	{

		public $db;

		private function startsWith($haystack, $needle) {
		    // search backwards starting from haystack length characters from the end
		    return $needle === "" || strrpos($haystack, $needle, -strlen($haystack)) !== false;
		}

		private function endsWith($haystack, $needle) {
		    // search forward starting from end minus needle length characters
		    return $needle === "" || (($temp = strlen($haystack) - strlen($needle)) >= 0 && strpos($haystack, $needle, $temp) !== false);
		}
		
		//Returns a DB with k:pairs v:wins-history.
		//Takes in the name of the formatted csv file.
		public function parseFile($file_name)
		{
			$ret = array();
			$read_state = 0; //0: looking for content, 1: in content
			$curr_arena_id = "";
			$curr_arena_wins = 0;

			$curr_arena_list = array();

			$contents = file_get_contents($file_name);
			$contents_array = explode("\n", $contents);

			foreach($contents_array as $line)
			{
				if($read_state == 0 && $this->startsWith($line, "game_version"))
					$read_state = 1;
				else if($read_state == 1)
				{
					//--if end of line
					if($this->startsWith($line, ",") || $this->startsWith($line, " "))
						break;

					$cols = explode(",", $line);

					//--if start of arena 'i'
					if($curr_arena_id != ($cols[1] . $cols[2] . $cols[6]))
					{
						//--do stuff with info from arena 'i-1'
						$curr_arena_pairs = $this->makeUniquePairs($curr_arena_list);
						for($i = 0; $i < count($curr_arena_pairs); $i++)
						{
							$pair = $curr_arena_pairs[$i];
							if(isset($ret[$pair->getHash()]))
							{
								$ret[$pair->getHash()]->history[] = $curr_arena_wins;
							}
							else
							{
								$ret[$pair->getHash()] = $pair;
								$ret[$pair->getHash()]->history = array();
								$ret[$pair->getHash()]->history[] = $curr_arena_wins;
							}
						}
						//echo "[Library] Arena $curr_arena_id has been added in. (" . count($ret) . " rows in db)<br>";

						//--prepare for arena 'i' like reset $curr_arena_id and add character
						$curr_arena_id = ($cols[1] . $cols[2] . $cols[6]);
						$curr_arena_wins = $cols[5];
						$curr_arena_list = array();
						$curr_arena_list[] = $cols[3];
					}

					//--(else) add card into $curr_arena_list
					$curr_arena_list[] = $cols[4];

				}	
			}

			$this->db = $ret;
			return $ret;
		}

		public function getScore($reference_pair)
		{
			$ret = 0;
			$db = $this->db;
			$pair = $db[$reference_pair->getHash()];
			$history = $pair->history;
			foreach($history as $score)
			{
				$ret += $score;
			}

			return ($ret/count($history));
		}

		public function getAllNextScores($history)
		{
			$ret = array();
			$list = $this->getUniqueCards();

			foreach($list as $card)
			{
				$historyPlus = $history;
				$historyPlus[] = $card;
				$pairsPlus = $this->makePairs($historyPlus);

				$cardScore = 0;
				foreach($pairsPlus as $pair)
				{
					if(isset($this->db[$pair->getHash()]))
						$cardScore += $this->getScore($pair);
					else
						$cardScore += 3;	//pair not in db
				}

				$ret[$card] = $cardScore;
			}

			return $ret;
		}


		private function makeUniquePairs($list)
		{
			$ret = array();

			for($i = 0; $i < count($list); $i++)
			{
				for($j = 0; $j < $i; $j++)
				{
					$ret[] = new Pair($list[$i], $list[$j], array());
				}
			}

			return $ret;
		}

		private function makePairs($list)
		{
			$ret = array();

			for($i = 0; $i < count($list); $i++)
			{
				for($j = 0; $j < count($list); $j++)
				{
					if($i != $j)
					{
						$currPair = new Pair($list[$i], $list[$j], array());
						$ret[] = $currPair;
					}
				}
			}

			return $ret;
		}

		private function getUniqueCards()
		{
			$ret = array();
			foreach($this->db as $pair)
			{
				if(!in_array($pair->a, $this->db)) $ret[] = $pair->a; 
				if(!in_array($pair->b, $this->db)) $ret[] = $pair->b;
			}

			return $ret;
		}


		



	}


?>